File: src/main/java/xyz/realmath/SubSpace.java
package xyz.realmath;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import com.google.errorprone.annotations.Immutable;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

@Immutable
public final class SubSpace {
  private final int embeddingDim;

  @SuppressWarnings("Immutable") // only written in constructor
  private final Map<Tuple<Rational>, Integer> basis;

  private SubSpace(int embeddingDim, Map<Tuple<Rational>, Integer> basis) {
    checkArgument(embeddingDim >= 0, "embeddingDim < 0");
    this.embeddingDim = embeddingDim;
    this.basis = new HashMap<>(basis);
  }

  @SafeVarargs
  public static SubSpace span(Tuple<Rational>... vectors) {
    checkArgument(vectors.length > 0, "require at least one vector");
    SubSpace retVal = zeroDim(vectors[0].dim());
    retVal = retVal.spanWith(vectors[0]);
    for (Tuple<Rational> v : vectors) {
      checkNotNull(v, "vector");
      checkArgument(v.dim() == retVal.embeddingDim, "vectors with different dim");
      retVal = retVal.spanWith(v);
    }
    return retVal;
  }

  public static SubSpace zeroDim(int embeddingDim) {
    return new SubSpace(embeddingDim, Collections.emptyMap());
  }

  public Set<Tuple<Rational>> basis() {
    return Collections.unmodifiableSet(basis.keySet());
  }

  public boolean contains(Tuple<Rational> vector) {
    checkArgument(vector.dim() == embeddingDim, "vector.length != embedding dim");
    // TODO: ZERO
    return reduce(vector).equals(vector.multiply(Rational.ZERO));
  }

  private Tuple<Rational> reduce(Tuple<Rational> vector) {
    for (Map.Entry<Tuple<Rational>, Integer> entry : basis.entrySet()) {
      Integer idx = entry.getValue();
      Rational v = vector.get(idx);
      if (!v.equals(Rational.ZERO)) {
        vector = entry.getKey().multiply(v).multiply(Rational.ONE.negate()).add(vector);
      }
    }
    return vector;
  }

  private SubSpace spanWith(Tuple<Rational> vector) {
    vector = reduce(vector);

    // normalize vector
    int i = 0;
    for (Rational r : vector) {
      if (!r.equals(Rational.ZERO)) {
        vector = vector.multiply(r.reciprocal());
        break;
      }
      i++;
    }
    if (i == embeddingDim) {
      return this;
    }

    // normalize old base vectors
    Map<Tuple<Rational>, Integer> newBasis = new HashMap<>();
    newBasis.put(vector, i);

    for (Map.Entry<Tuple<Rational>, Integer> entry : basis.entrySet()) {
      Tuple<Rational> t1 = entry.getKey();
      Rational v = entry.getKey().get(i);
      if (!v.equals(Rational.ZERO)) {
        t1 = vector.multiply(v).multiply(Rational.ONE.negate()).add(entry.getKey());
      }
      newBasis.put(t1, entry.getValue());
    }
    return new SubSpace(embeddingDim, newBasis);
  }

  public int dim() {
    return basis.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null || getClass() != obj.getClass()) {
      return false;
    }
    SubSpace that = (SubSpace) obj;
    if (that.embeddingDim != embeddingDim) {
      return false;
    }

    for (Tuple<Rational> vector : basis.keySet()) {
      if (!that.contains(vector)) {
        return false;
      }
    }

    for (Tuple<Rational> vector : that.basis.keySet()) {
      if (!contains(vector)) {
        return false;
      }
    }

    return true;
  }

  @Override
  public int hashCode() {
    return Objects.hash(embeddingDim, dim());
  }
}

File: src/main/kotlin/xyz/realmath/SubSpace.kt
package xyz.realmath

class SubSpace private constructor(private val embeddingDim: Int, private val basis: Map<Tuple<Rational>, Int>) {

    val dim = basis.size

    init {
        check(embeddingDim >= 0) { "embeddingDim < 0" }
    }

    fun basis(): Set<Tuple<Rational>> = basis.keys

    operator fun contains(vector: Tuple<Rational>): Boolean {
        check(vector.dim == embeddingDim) { "vector.length != embedding dim" }
        // TODO: ZERO
        return reduce(vector) == vector as Tuple<out Rational> * Rational.ZERO
    }

    private fun reduce(vector: Tuple<Rational>): Tuple<Rational> {
        var retVal = vector
        basis.forEach { key, idx ->
            val v = retVal[idx]
            if (v != Rational.ZERO) {
                retVal -= key as Tuple<out Rational> * v
            }
        }
        return retVal
    }

    private fun spanWith(vector: Tuple<Rational>): SubSpace {
        var tmpVector = reduce(vector)

        // normalize vector
        var i = 0
        for (r in tmpVector) {
            if (r != Rational.ZERO) {
                tmpVector /= r
                break
            }
            i++
        }
        if (i == embeddingDim) {
            return this
        }

        // normalize old base vectors
        val newBasis = HashMap<Tuple<Rational>, Int>()
        newBasis.put(tmpVector, i)

        for (entry in basis.entries) {
            var t1 = entry.key
            val v = entry.key[i]
            if (v != Rational.ZERO) {
                t1 -= tmpVector as Tuple<out Rational> * v
            }
            newBasis.put(t1, entry.value)
        }
        return SubSpace(embeddingDim, newBasis)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is SubSpace) {
            return false
        }
        if (other.embeddingDim != embeddingDim) {
            return false
        }

        return basis.keys.all { other.contains(it) } && other.basis.keys.all { contains(it) }
    }

    override fun hashCode(): Int {
        var result = embeddingDim
        result = 31 * result + dim
        return result
    }

    companion object {
        fun span(vararg vectors: Tuple<Rational>): SubSpace {
            check(vectors.size > 0) { "require at least one vector" }
            var retVal = zeroDim(vectors[0].dim)
            retVal = retVal.spanWith(vectors[0])
            for (v in vectors) {
                checkNotNull(v) { "vector" }
                check(v.dim == retVal.embeddingDim) { "vectors with different dim" }
                retVal = retVal.spanWith(v)
            }
            return retVal
        }

        fun zeroDim(embeddingDim: Int): SubSpace {
            return SubSpace(embeddingDim, emptyMap())
        }
    }
}
